Retain cursorlineopt=number in terminal mode

yorickpeterse opened the pull request:
When entering terminal mode, cursorlineopt is no longer entirely
disabled. Instead, it's set to `number`. Doing so ensures that users
using `set cursorline` combined with `set cursorlineopt=number` have
consistent highlighting of the line numbers, instead of this being
disabled when entering terminal mode.

The patch was suggested by Gregory Anders in comment
https://github.com/neovim/neovim/issues/15481#issuecomment-905742792.

This fixes https://github.com/neovim/neovim/issues/15481

YorickPeterse commented on src/nvim/terminal.c:
```diff
@@ -359,11 +359,13 @@ void terminal_enter(void)
   long save_w_p_so = curwin->w_p_so;
   long save_w_p_siso = curwin->w_p_siso;
-  curwin->w_p_cul = false;
+  curwin->w_p_culopt = (char_u *)xstrdup("number");
```
@seandewar Maybe you know more that can help: ASAN complains this line isn't free'd, in spite of the `xfree` further below. I'm guessing `save_curwin != curwin`, preventing the `xfree` from running. I thought about moving the `xfree` out of the `if` and use `xfree(save_curwin->w_p_culopt);`, but I'm not sure if `save_curwin` is still valid at that point. Do you have any suggestions on how to approach this?

YorickPeterse replied:
Log output in question:

```
2021-08-26T20:26:33.4694282Z = ==32671==ERROR: LeakSanitizer: detected memory leaks
2021-08-26T20:26:33.4694894Z = 
2021-08-26T20:26:33.4695483Z = Direct leak of 7 byte(s) in 1 object(s) allocated from:
2021-08-26T20:26:33.4696117Z =     #0 0x4d8f1d in malloc ??:?
2021-08-26T20:26:33.4696833Z =     #1 0x10c7f12 in try_malloc /home/runner/work/neovim/neovim/src/nvim/memory.c:73:15
2021-08-26T20:26:33.4697698Z =     #2 0x10c810c in xmalloc /home/runner/work/neovim/neovim/src/nvim/memory.c:107:15
2021-08-26T20:26:33.4698574Z =     #3 0x10c86a6 in xmallocz /home/runner/work/neovim/neovim/src/nvim/memory.c:184:15
2021-08-26T20:26:33.4699457Z =     #4 0x10c87fe in xmemdupz /home/runner/work/neovim/neovim/src/nvim/memory.c:202:17
2021-08-26T20:26:33.4700598Z =     #5 0x10c9c08 in xstrdup /home/runner/work/neovim/neovim/src/nvim/memory.c:421:10
2021-08-26T20:26:33.4701516Z =     #6 0x192a666 in terminal_enter /home/runner/work/neovim/neovim/src/nvim/terminal.c:367:34
2021-08-26T20:26:33.4702404Z =     #7 0x849ee8 in edit /home/runner/work/neovim/neovim/src/nvim/edit.c:1383:7
2021-08-26T20:26:33.4703249Z =     #8 0x1230d8c in invoke_edit /home/runner/work/neovim/neovim/src/nvim/normal.c:7752:7
2021-08-26T20:26:33.4704123Z =     #9 0x11fbe52 in nv_edit /home/runner/work/neovim/neovim/src/nvim/normal.c:7724:5
2021-08-26T20:26:33.4705006Z =     #10 0x11e4c58 in normal_execute /home/runner/work/neovim/neovim/src/nvim/normal.c:1147:3
2021-08-26T20:26:33.4705892Z =     #11 0x1845779 in state_enter /home/runner/work/neovim/neovim/src/nvim/state.c:69:26
2021-08-26T20:26:33.4706773Z =     #12 0x11a32fa in normal_enter /home/runner/work/neovim/neovim/src/nvim/normal.c:463:3
2021-08-26T20:26:33.4707622Z =     #13 0xf4fe7d in main /home/runner/work/neovim/neovim/src/nvim/main.c:560:3
2021-08-26T20:26:33.4709327Z =     #14 0x7f0696dee0b2 in __libc_start_main /build/glibc-eX1tMB/glibc-2.31/csu/../csu/libc-start.c:308:16
2021-08-26T20:26:33.4710115Z = 
2021-08-26T20:26:33.4710693Z = Direct leak of 5 byte(s) in 1 object(s) allocated from:
2021-08-26T20:26:33.4711338Z =     #0 0x4d8f1d in malloc ??:?
2021-08-26T20:26:33.4712049Z =     #1 0x10c7f12 in try_malloc /home/runner/work/neovim/neovim/src/nvim/memory.c:73:15
2021-08-26T20:26:33.4712922Z =     #2 0x10c810c in xmalloc /home/runner/work/neovim/neovim/src/nvim/memory.c:107:15
2021-08-26T20:26:33.4713795Z =     #3 0x10c86a6 in xmallocz /home/runner/work/neovim/neovim/src/nvim/memory.c:184:15
2021-08-26T20:26:33.4715278Z =     #4 0x10c87fe in xmemdupz /home/runner/work/neovim/neovim/src/nvim/memory.c:202:17
2021-08-26T20:26:33.4716171Z =     #5 0x10c9c08 in xstrdup /home/runner/work/neovim/neovim/src/nvim/memory.c:421:10
2021-08-26T20:26:33.4717019Z =     #6 0x1847cfb in vim_strsave /home/runner/work/neovim/neovim/src/nvim/strings.c:58:20
2021-08-26T20:26:33.4717901Z =     #7 0x1323ad3 in set_string_option_direct /home/runner/work/neovim/neovim/src/nvim/option.c:2165:7
2021-08-26T20:26:33.4718798Z =     #8 0x12ddf86 in set_option_default /home/runner/work/neovim/neovim/src/nvim/option.c:620:9
2021-08-26T20:26:33.4719692Z =     #9 0x12d703a in set_options_default /home/runner/work/neovim/neovim/src/nvim/option.c:679:7
2021-08-26T20:26:33.4720537Z =     #10 0x12d4ab8 in set_init_1 /home/runner/work/neovim/neovim/src/nvim/option.c:525:3
2021-08-26T20:26:33.4721328Z =     #11 0xf4cff7 in early_init /home/runner/work/neovim/neovim/src/nvim/main.c:194:3
2021-08-26T20:26:33.4722108Z =     #12 0xf4dc8b in main /home/runner/work/neovim/neovim/src/nvim/main.c:247:3
2021-08-26T20:26:33.4723276Z =     #13 0x7f0696dee0b2 in __libc_start_main /build/glibc-eX1tMB/glibc-2.31/csu/../csu/libc-start.c:308:16
2021-08-26T20:26:33.4723965Z = 
2021-08-26T20:26:33.4724586Z = SUMMARY: AddressSanitizer: 12 byte(s) leaked in 2 allocation(s).
2021-08-26T20:26:33.4725246Z ====================================================================================================
2021-08-26T20:26:33.4726290Z [2mnan ms[0m test/helpers.lua:249: Found runtime errors in logfile(s): /home/runner/work/neovim/neovim/build/log/ubsan.32671
```

YorickPeterse replied:
Per Matrix:

```
gpanders@gpanders:matrix.org
seems to work fine just now when I tested it
xstrdup just allocates a new copy of its argument
but you should be able to just do (char_u *)"number"
```

I shall give this a try.

YorickPeterse commented:
Finally all the tests are green. 

![300px-But_It's_Honest_Work](https://user-images.githubusercontent.com/86065/131919274-9ed372d6-7d50-4a5d-9557-38befb3bda8f.jpg)

Thanks for the help @seandewar @gpanders!

gpanders commented on src/nvim/terminal.c:
```diff
@@ -386,10 +388,14 @@ void terminal_enter(void)
     curwin->w_p_cuc = save_w_p_cuc;
     curwin->w_p_so = save_w_p_so;
     curwin->w_p_siso = save_w_p_siso;
+  } else {
+      xfree(save_w_p_culopt);
```
```suggestion
  } else {
    xfree(save_w_p_culopt);
```

gpanders commented:
We should probably check the user's current value of `'cursorlineopt'` and only set it to "number" if the current value is either "number" or "both".

gpanders commented on src/nvim/terminal.c:
```diff
@@ -359,11 +359,13 @@ void terminal_enter(void)
   // placed at end of buffer to "follow" output. #11072
   win_T *save_curwin = curwin;
-  int save_w_p_cul = curwin->w_p_cul;
+  char_u *save_w_p_culopt = curwin->w_p_culopt;
+  char_u save_w_p_culopt_flags = curwin->w_p_culopt_flags;
```
See comment below.

```suggestion
  char_u *save_w_p_culopt = NULL;
  char_u save_w_p_culopt_flags = curwin->w_p_culopt_flags;
```

gpanders commented on src/nvim/terminal.c:
```diff
@@ -359,11 +359,13 @@ void terminal_enter(void)
   // Disable these options in terminal-mode. They are nonsense because cursor is
   // placed at end of buffer to "follow" output. #11072
   win_T *save_curwin = curwin;
-  int save_w_p_cul = curwin->w_p_cul;
```
Let's keep this line, since if `'cursorlineopt'` equals e.g. `"line"` then we need to disable `cursorline` altogether. We will just restore it later in any case. (Although you can change `int` to `bool`).

gpanders commented on src/nvim/terminal.c:
```diff
@@ -359,11 +359,13 @@ void terminal_enter(void)
   int save_w_p_cuc = curwin->w_p_cuc;
   long save_w_p_so = curwin->w_p_so;
   long save_w_p_siso = curwin->w_p_siso;
-  curwin->w_p_cul = false;
+  curwin->w_p_culopt = (char_u *)xstrdup("number");
+  curwin->w_p_culopt_flags = CULOPT_NBR;
```
Let's check the current value of `curwin->w_p_culopt_flags` to make sure `"number"` is already present in the user's setting. If it's not (e.g. `cursorlineopt=line`) then we need to disable `'cursorline'`.

Also if the user's value of `'cursorlineopt'` is already just "number" we can avoid the string allocation.

```suggestion
  if (curwin->w_p_culopt_flags & CULOPT_NBR) {
    if (strcmp(curwin->w_p_culopt, "number")) {
        save_w_p_culopt = curwin->w_p_culopt;
        curwin->w_p_culopt = (char_u *)xstrdup("number");
    }
    curwin->w_p_culopt_flags = CULOPT_NBR;
  } else {
    curwin->w_p_cul = false;
  }
```

gpanders commented on src/nvim/terminal.c:
```diff
@@ -386,10 +388,14 @@ void terminal_enter(void)
 
   if (save_curwin == curwin) {  // save_curwin may be invalid (window closed)!
-    curwin->w_p_cul = save_w_p_cul;
+    xfree(curwin->w_p_culopt);
+    curwin->w_p_culopt_flags = save_w_p_culopt_flags;
+    curwin->w_p_culopt = save_w_p_culopt;
     curwin->w_p_cuc = save_w_p_cuc;
     curwin->w_p_so = save_w_p_so;
     curwin->w_p_siso = save_w_p_siso;
+  } else {
+    xfree(save_w_p_culopt);
```
```suggestion
    curwin->w_p_cul = save_w_p_cul;
    if (save_w_p_culopt) {
      xfree(curwin->w_p_culopt);
      curwin->w_p_culopt = save_w_p_culopt;
    }
    curwin->w_p_culopt_flags = save_w_p_culopt_flags;
    curwin->w_p_cuc = save_w_p_cuc;
    curwin->w_p_so = save_w_p_so;
    curwin->w_p_siso = save_w_p_siso;
  } else if (save_w_p_culopt) {
    xfree(save_w_p_culopt);
```

gpanders commented on test/functional/terminal/buffer_spec.lua:
```diff
@@ -22,14 +22,14 @@ describe(':terminal buffer', function()
     feed('i')
     eq('t', eval('mode()'))
-    eq({ 0, 0, 0, 0 }, eval('[&l:cursorline, &l:cursorcolumn, &l:scrolloff, &l:sidescrolloff]'))
+    eq({ 'number', 1, 0, 0, 0 }, eval('[&l:cursorlineopt, &l:cursorline, &l:cursorcolumn, &l:scrolloff, &l:sidescrolloff]'))
```
Let's add a test case here for `cursorlineopt=line` and make sure `cursorline` is still disabled.

YorickPeterse commented:
Just as an FYI: I'll be busy in the coming days, so I'm not sure when I'll have time to look at this PR again. If somebody wants to take over that might speed things up :smiley:

yorickpeterse committed:
fix: set cursorlineopt=number in terminal mode

When entering terminal mode, cursorlineopt is no longer entirely
disabled. Instead, it's set to `number`. Doing so ensures that users
using `set cursorline` combined with `set cursorlineopt=number` have
consistent highlighting of the line numbers, instead of this being
disabled when entering terminal mode.

The patch was suggested by Gregory Anders in comment
https://github.com/neovim/neovim/issues/15481#issuecomment-905742792.

This fixes https://github.com/neovim/neovim/issues/15481

Co-Authored-By: Gregory Anders <greg@gpanders.com>
Co-Authored-By: Sean Dewar <seandewar@users.noreply.github.com>
